http/server.go

```diff
@@ -13,304 +13,171 @@ package http
-
-		if s.options.TLS.Value().Email == "" || s.options.TLS.Value().Email == config.DefaultAPIEmail {
-			return ErrNoEmail
-		}
-
-		certCache := path.Join(s.options.RootDir, "autocerts")
-
-		log.FeedbackInfo(
-			ctx,
-			"Generating auto certificate",
-			logging.NewKV("Domain", s.options.Domain.Value()),
-			logging.NewKV("Certificate cache", certCache),
-		)
-
-		m := &autocert.Manager{
-			Cache:      autocert.DirCache(certCache),
-			Prompt:     autocert.AcceptTOS,
-			Email:      s.options.TLS.Value().Email,
-			HostPolicy: autocert.HostWhitelist(s.options.Domain.Value()),
-		}
-
-		cfg.GetCertificate = m.GetCertificate
-
-		// We set manager on the server instance to later start
-		// a redirection server.
-		s.certManager = m
-	} else {
-		// When not using auto cert, we create a self signed certificate
-		// with the provided public and prive keys.
-		log.FeedbackInfo(ctx, "Generating self signed certificate")
-
-		cert, err := tls.LoadX509KeyPair(
-			s.options.TLS.Value().PrivateKey,
-			s.options.TLS.Value().PublicKey,
-		)
-		if err != nil {
-			return NewErrFailedToLoadKeys(err, s.options.TLS.Value().PublicKey, s.options.TLS.Value().PrivateKey)
-		}
-
-		cfg.Certificates = []tls.Certificate{cert}
-	}
-
-	var err error
-	s.listener, err = tls.Listen("tcp", s.Addr, cfg)
-	if err != nil {
-		return errors.WithStack(err)
-	}
-
-	// Save the address on the server in case the port was set to random
-	// and that we want to see what was assigned.
-	s.address = s.listener.Addr().String()
-
-	return nil
+// Server struct holds the Handler for the HTTP API.
+type Server struct {
+	address atomic.Value
+	options *ServerOptions
+	server  *http.Server
 }
 
-// Run calls Serve with the receiver's listener.
-func (s *Server) Run(ctx context.Context) error {
-	if s.listener == nil {
-		return ErrNoListener
+// NewServer instantiates a new server with the given http.Handler.
+func NewServer(handler http.Handler, opts ...ServerOpt) (*Server, error) {
+	options := DefaultServerOptions()
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	// setup a mux with the default middleware stack
+	mux := chi.NewMux()
+	mux.Use(
+		middleware.RequestLogger(&logFormatter{}),
+		middleware.Recoverer,
+		CorsMiddleware(options.AllowedOrigins),
+	)
+	mux.Handle("/*", handler)
+
+	server := &http.Server{
+		ReadTimeout:  options.ReadTimeout,
+		WriteTimeout: options.WriteTimeout,
+		IdleTimeout:  options.IdleTimeout,
+		TLSConfig:    tlsConfig,
+		Handler:      mux,
+	}
+
+	var address atomic.Value
+	address.Store("")
+
+	return &Server{
+		address: address,
+		options: options,
+		server:  server,
+	}, nil
+}
+
+// ListenAndServe listens for and serves incoming connections.
+func (s *Server) ListenAndServe() error {
+	listener, err := net.Listen("tcp", s.options.Address)
+	if err != nil {
+		return err
 	}
+	// ignore close errors as they cannot be handled
+	// from the caller of this method
+	defer listener.Close() //nolint:errcheck
```

Comment:
question: Why can they not be handled but other errors appear to be so?  And is that not up to any callers of this public function to decide?  Should the choice to discard the error not be made by the caller?

Reply:
I've removed this `Close` call as it is redundant. After taking a closer look the `Serve` and `ServeTLS` functions will close the listener.